prow/tide/history/history.go

```diff
@@ -37,8 +47,95 @@ var now = time.Now
 type History struct {
 	logs map[string]*recordLog
 	sync.Mutex
-
 	logSizeLimit int
+
+	gcs *GCS
+}
+
+type GCS struct {
+	prefix string
+	bucket *storage.BucketHandle
+}
+
+func NewGCS(gcsCredentialsFile, bucket, prefix string) (*GCS, error) {
+	client, err := storage.NewClient(context.Background(), option.WithCredentialsFile(gcsCredentialsFile))
+	if err != nil {
+		return nil, fmt.Errorf("error creating GCS client: %v", err)
+	}
+	return &GCS{
+		prefix: prefix,
+		bucket: client.Bucket(bucket),
+	}, nil
+}
+
+func (g *GCS) readHistory() (map[string][]Record, error) {
+	ctx := context.Background()
+	objIter := g.bucket.Objects(ctx, &storage.Query{Prefix: g.prefix})
+	res := make(map[string][]Record)
+	for {
+		objAttr, err := objIter.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, fmt.Errorf("error iterating GCS objects with prefix %q: %v", g.prefix, err)
+		}
+		reader, err := g.bucket.Object(objAttr.Name).NewReader(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("error opening GCS object %q: %v", objAttr.Name, err)
+		}
+		defer reader.Close()
+		b, err := ioutil.ReadAll(reader)
+		if err != nil {
+			return nil, fmt.Errorf("error reading GCS object %q: %v", objAttr.Name, err)
+		}
+		var recs []Record
+		if err := json.Unmarshal(b, &recs); err != nil {
+			return nil, fmt.Errorf("error unmarshaling GCS object %q: %v", objAttr.Name, err)
+		}
+		poolKey := strings.TrimPrefix(objAttr.Name, g.prefix)
+		res[poolKey] = recs
+	}
+	logrus.Debugf("Successfully read action history for %d pools.", len(res))
+	return res, nil
+}
+
+func (g *GCS) writeHistory(hist map[string][]*Record) error {
```

Comment:
It certainly isn't necessary, I just figured it would make the GCS data easier for a human to inspect. In retrospect I'm not sure how useful that is. 
I can change this to write one GCS object (`History.logs` serialized) if that is preferred.

Reply:
I think this would be better, because then we don't have to worry about cleaning up old entries, etc.

Reply:
Done.